src/lava/proc/sparse/models.py

```diff
@@ -0,0 +1,359 @@
+            self.weights = truncate_weights(
+                self.weights,
+                sign_mode=self.sign_mode,
+                num_weight_bits=self.num_weight_bits
+            )
+            self.weights_set = True
+
+        # The a_out sent at each timestep is a buffered value from dendritic
+        # accumulation at timestep t-1. This prevents deadlocking in
+        # networks with recurrent connectivity structures.
+        self.a_out.send(self.a_buff)
+        if self.num_message_bits.item() > 0:
+            s_in = self.s_in.recv()
+            a_accum = self.weights.dot(s_in)
+        else:
+            s_in = self.s_in.recv().astype(bool)
+            a_accum = self.weights[:, s_in].sum(axis=1).A1
+
+        self.a_buff = (
+            np.left_shift(a_accum, self.weight_exp)
+            if self.weight_exp > 0
+            else np.right_shift(a_accum, -self.weight_exp)
+        )
+
+        self.recv_traces(s_in)
+
+
+class AbstractPyDelaySparseModel(PyLoihiProcessModel):
+    """Abstract Conn Process with Sparse synaptic connections which incorporates
+    delays into the Conn Process.
+    """
+
+    @staticmethod
+    def get_del_wgts(weights, delays) -> spmatrix:
+        """
+        Use self.weights and self.delays to create a matrix where the
+        weights are separated by delay. Returns 2D matrix of form
+        (num_flat_output_neurons * max_delay + 1, num_flat_input_neurons) where
+        del_wgts[
+            k * num_flat_output_neurons : (k + 1) * num_flat_output_neurons, :
+        ]
+        contains the weights for all connections with a delay equal to k.
+        This allows for the updating of the activation buffer and updating
+        weights.
+        """
+        # Can only start at 1, as delays==0 raises inefficiency warning
+        weight_delay_from_1 = vstack([weights.multiply(delays == k)
+                                      for k in range(1, np.max(delays) + 1)])
+        # Create weight matrix at delays == 0
+        r, c, _ = find(delays)
+        weight_delay_zeros = weights.copy()
+        weight_delay_zeros[r, c] = 0
+        weight_delay_zeros.eliminate_zeros()
+        return vstack([weight_delay_zeros, weight_delay_from_1])
+
+    def calc_act(self, s_in) -> np.ndarray:
+        """
+        Calculate the activations by performing del_wgts * s_in. This matrix
+        is then summed across each row to get the activations to the output
+        neurons for different delays. This activation vector is reshaped to a
+        matrix of the form
+        (n_flat_output_neurons * (max_delay + 1), n_flat_output_neurons)
+        which is then transposed to get the activation matrix.
+        """
+        return np.reshape(self.get_del_wgts(self.weights,
+                                            self.delays).dot(s_in),
+                          (np.max(self.delays) + 1,
+                          self.weights.shape[0])).T
+
+    def update_act(self, s_in):
```

Comment:
If this is the main method, put it at the top of the class and put the other ones, that it uses, below.

Reply:
done